src/Sulu/Bundle/ContentBundle/Serializer/Subscriber/StructureSubscriber.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+/*
+ * This file is part of Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Bundle\ContentBundle\Serializer\Subscriber;
+
+use JMS\Serializer\EventDispatcher\Events;
+use JMS\Serializer\EventDispatcher\EventSubscriberInterface;
+use JMS\Serializer\EventDispatcher\ObjectEvent;
+use JMS\Serializer\EventDispatcher\PreSerializeEvent;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\DocumentInspector;
+use Sulu\Component\Content\Document\Behavior\StructureBehavior;
+use Sulu\Component\Content\Document\Structure\Structure;
+use Sulu\Component\Content\Metadata\StructureMetadata;
+use Sulu\Component\DocumentManager\Behavior\Mapping\UuidBehavior;
+
+/**
+ * Normalize ManagedStructure instances to the Structure type.
+ */
+class StructureSubscriber implements EventSubscriberInterface
+{
+    /**
+     * @var DocumentInspector
+     */
+    private $inspector;
+
+    public function __construct(DocumentInspector $inspector)
+    {
+        $this->inspector = $inspector;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return [
+            [
+                'event' => Events::PRE_SERIALIZE,
+                'method' => 'onPreSerialize',
+            ],
+            [
+                'event' => Events::POST_SERIALIZE,
+                'format' => 'json',
+                'method' => 'onPostSerialize',
+            ],
+        ];
+    }
+
+    /**
+     * @param PreSerializeEvent $event
+     */
+    public function onPreSerialize(PreSerializeEvent $event)
+    {
+        if ($event->getObject() instanceof Structure) {
+            $event->setType(Structure::class);
+        }
+    }
+
+    public function onPostSerialize(ObjectEvent $event)
+    {
+        $document = $event->getObject();
+        $context = $event->getContext();
+
+        if (!$document instanceof StructureBehavior) {
+            return;
+        }
+
+        $visitor = $event->getVisitor();
+
+        $visitor->addData('template', $document->getStructureType());
+        $visitor->addData('originTemplate', $document->getStructureType());
+        $visitor->addData('internal', false);
+
+        if (array_search('defaultPage', $context->attributes->get('groups')->getOrElse([])) !== false) {
+            $structure = $document->getStructure();
```

Comment:
TBH, I am not sure... That's the biggest issue with the current structure, there are so many classes doing essentially the same, with similar names, but without a common interface that can be used. That's something we should improve step by step.

PS: I just debugged it, it uses a `ManagedStructure` (so there is even a third class :confused:)